mypy/checkexpr.py

```diff
@@ -2520,7 +2520,18 @@ def nonliteral_tuple_index_helper(self, left_type: TupleType, index: Expression)
                                       'actual type', 'expected type'):
             return AnyType(TypeOfAny.from_error)
         else:
-            return UnionType.make_simplified_union(left_type.items)
+            union = UnionType.make_simplified_union(left_type.items)
+            # If we are indexing, we return the union of elements. In a slice,
+            # we return Tuple[Any, ...], unless it is homogeneous, where we return
+            # Tuple[T, ...]
+            if isinstance(index, SliceExpr):
+                if union == left_type.items[0]:
+                    t = union
+                else:
+                    t = AnyType(TypeOfAny.implementation_artifact)
```

Comment:
I think that it would be better to produce `Tuple[Union[...], ...]` instead of `Tuple[Any, ...]`. The rationale is that the bug has been around for quite a long time so the rate of false positives is expected to low. On the other hand, producing `Any` can produce imprecise typing in an otherwise fully typed program.

Reply:
Ok, then I will do that.